### ISSUE TITLE
Put minibatch work onto master

### ISSUE THREAD

**Issue by alexggmatthews** [GPflow/GPflow#7]
We have done some work on getting minibatches working in GPflow. We should consolidate it into its simplest and cleanest form and then submit a pull request onto master.

**Comment by markvdw** [comment]
Hi Alex, I'm happy to do this today.

**Comment by alexggmatthews** [comment]
OK Mark great! We're trying to keep 1 issue per pull request. Then we can review it there.

**Comment by markvdw** [comment]
So I have an implementation running, but before I code things together, I wanted a bit of input, in particular on:
- Which class(es) should be SVI "aware"?
- Should we add compatibility for the Tensorflow optimisers?
These two are related.

There are two ways of adding the SVI functionality:

- Inside classes that inherit from `Model`. For example, we want to add SVI to `SVGP`. In this case, we modify `SVGP.build_likelihood` to randomly subsample `self.X`, calculate the likelihood and rescale appropriately. Advantage: `Model` stays untouched.
- Inside `Model`. In this case, we make `self.X` a Tensorflow variable, which is passed into the computational graph when `self._session.run()` is called. `self.objective` could be rewritten to subsample the data if a particular flag in `Model` was set. Advantage: In the Tensorflow docs, this is the way minibatches are handled. Disadvantage: `Model` needs to be SVI aware.

Personally, I would argue for putting the SVI functionality in `Model`, for the following reasons:
- The subsampling gets done whenever the objective is run, depending on what dataset is passed into the computational graph. If the subsampling is done in `build_likelihood`, the model would need to be recompiled each time the minibatch size is changed.
- The above goes hand-in-hand with what's needed to use the Tensorflow way of handling stochastic optimisation.
- `Model` contains the optimisation procedure, which needs to be SVI aware.
- This change can be transparent to models which do not use minibatches. For models that don't support minibatches, nothing changes so long as the full dataset is passed into the computational graph. Even if the full dataset isn't passed, nothing will break (though we won't get an unbiased estimate of the objective).

An alternative would be to create another intermediate class `SVIModel` or something, from which other models which require SVI would inherit. This class would have to re-implement some stuff that `Model` does with only minor changes (which looks like code duplication to me).

The second issue, whether to support Tensorflow optimisers, is arguably orthogonal to stochastic estimation of the bound. But what use is estimating a bound if you don't optimise it? So currently I have Tensorflow optimisers running fine, but it required a bit of a large change to `Model.optimize()`. If Tensorflow optimisers are to be supported a separate code path is needed for running scipy optimisers to the Tensorflow ones. Currently I did this by essentially renaming the current `optimize` routine `Model._np_optimize`, which gets called if the `method` argument is a string. If `method` is a Tensorflow optimiser step, then `Model._tf_optimize` is called. The code would be fairly clean if it weren't for a hack requiring casting the parameters and objective function outputs to `float32` when using Tensorflow optimisers, due to https://github.com/tensorflow/tensorflow/issues/761.

So the question here is, should we go down the Tensorflow optimisation route? Should this be in the same pull request?

Again I would argue for it being in the same pull request, since implementing SVI is only really done with a working optimisation routine.

Thoughts are welcome!

**Comment by markvdw** [comment]
Actually, on reflection, things are a bit more complicated since `Model` doesn't have a concept of the training dataset (`self.X` and `self.Y`). When using the Tensorflow optimisers, these have to be Variables part of the computational graph, and need to be passed using `feed_dict` when running the optimisation step.

I'm now thinking that it would be best for `GPModel` to have a overridden `_compile()`, and for `Model._tf_optimize` to take as a parameter extra variable to add to `feed_dict`.

**Comment by markvdw** [comment]
Ok, so I've submitted two branches, `svi` and `svi-tfopt`. The first one has the minimal amount of modifications to allow stochastic evaluation of the bound, while `svi-tfopt` contains the modifications required to use the Tensorflow optimisers as well. I think it turned out pretty clean, but let me know any thoughts.

**Comment by alexggmatthews** [comment]
Hi Mark,  Great work!

I had a look at the svi branch. It looks good in principle but the unit tests currently fail. I doubt it will be very difficult to fix. In order to avoid your code getting log jammed I would be happy to merge this once we've dealt with the outstanding issues.

I also had a look at svi-tfopt . I think this is a really important thing to do. The unit tests currently fail here too. I think this is quiet a bit more complex and I think needs discussion with @jameshensman as well. 

Alex

**Comment by markvdw** [comment]
Updates of branches `svi` and `svi-tfopt` are available. I went for the solution of keeping `self.X` and `self.Y` numpy arrays which contain the data. The minibatch Tensorflow variables which are fed into the computational graph are now `self._tfX` and `self._tfY`.

All unittests now pass.

**Comment by jameshensman** [comment]
This looks really nice. 

I like the separation of the scipy optimizers from the tensorflow ones, really nice. 

I hope that at some point, we can remove the float32 trick (presumably when the tensorflow optimizers support 64 bit). 

Perhaps the minibatch setter should set _needs_recompile?

My only query is why we have to pass the X, Y data as part of the feed_dict. Wouldn't it be better to just select a subset of the data inside SVGP.build_likelihood? That's the only algorithm that can handle a sub set of data. 

Also, the passing of tf_X, tf_Y affects all the models now. But other than the svgp, the models still use self.X and self.Y, not self.tf_X, self.tf_Y. Affected:
gpr.py
vgp.py
gpmc.py
sgpmc.py

Suggestions: 
a) use tf_X, tf_Y everywhere. Or
b) use self.X, self.Y everywhere and choose a sub-set in svgp.py if needed (and so remove the _compile overwrite in the GPModel class).

**Comment by markvdw** [comment]
So the reason the datasets are passed using `feed_dict` is mostly because this is what's recommended in the Tensorflow docs. The advantage is that you don't need to load the full dataset into memory in one go. It could be read sequentially from a hard drive, or network etc.

I don't think the minibatch setter needs to set `_needs_recompile`, since the scaling of the objective function can be inferred from the size of `_tfX` it gets.

You're right about `_tfX` being passed to all the models now. The reason I still did it this way is because now it's possible for models inheriting from `GPModel` to choose whether they want to be influenced by the minibatch procedure or not. So currently, by using `self.X`, the whole minibatch procedure is bypassed. `self._tfX` is passed into the computational graph, but not used.

It may be more clear though to have all models using the same data variables. This could prevent bugs like "I set my minibatch for GPR, but I'm getting the same results!". I could possibly also add a variable which causes an exception to be thrown if `minibatch_size` is set in models that don't support this.

So what do you reckon, is it better to make all models use the X and Y fed in through `feed_dict`?

**Comment by jameshensman** [comment]
Okay, I decided to move the minibatching code the the svgp class for now, since that's the only class that can do minibatching anyway. @markvdw , if you're happy with my edits, let me know and I'll accept this PR.

**Comment by markvdw** [comment]
Yes, happy with this!

**Comment by jameshensman** [comment]
merged.
